// @sitecore-jss/sitecore-jss-rendering-host/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tunnel_1 = require("./tunnel");
Object.defineProperty(exports, "startRenderHostTunnel", { enumerable: true, get: function () { return tunnel_1.startRenderHostTunnel; } });
var devServer_1 = require("./devServer");
Object.defineProperty(exports, "startDevServer", { enumerable: true, get: function () { return devServer_1.startDevServer; } });
var ssrMiddleware_1 = require("./ssrMiddleware");
Object.defineProperty(exports, "ssrMiddleware", { enumerable: true, get: function () { return ssrMiddleware_1.ssrMiddleware; } });
var renderingHostServer_1 = require("./renderingHostServer");
Object.defineProperty(exports, "startRenderingHostServer", { enumerable: true, get: function () { return renderingHostServer_1.startRenderingHostServer; } });
var defaultAppInvocationInfoResolver_1 = require("./defaultAppInvocationInfoResolver");
Object.defineProperty(exports, "getDefaultAppInvocationInfoResolver", { enumerable: true, get: function () { return defaultAppInvocationInfoResolver_1.getDefaultAppInvocationInfoResolver; } });
// const jssConfig = require('../scjssconfig.json');
// const ssrWebpackConfig = require('../server/server.webpack.config');
// const craWebpackConfig = require('../node_modules/react-scripts/config/webpack.config');
// const craWebpackDevServerConfig = require('../node_modules/react-scripts/config/webpackDevServer.config');
// doIt(5000);
// // invoking code, to be called from within sample app
// function doIt(port = 5000) {
//   startRenderHostTunnel('http://localhost', { port })
//     .then((tunnelUrl: string) => {
//       // generateHtmlTemplate(tunnelUrl);
//       startDevServer({
//         port,
//         tunnelUrl,
//         configFactory,
//         buildArtifactsPath: '../build',
//         serverBundleFileName: 'server.bundle',
//         urlToOpenOnStart: `${jssConfig.sitecore.layoutServiceHost}?sc_renderengineurl=${tunnelUrl}`,
//       });
//     })
//     .catch((err: Error) => {
//       console.error(err);
//     });
// }
