// @sitecore-jss/sitecore-jss-cli/dist/scripts/deploy.component.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.args = void 0;
var chalk_1 = __importDefault(require("chalk"));
var micro_manifest_1 = __importDefault(require("../micro-manifest"));
var deploy_template_1 = require("./deploy.template");
/**
 * @param {Argv} yargs
 */
function builder(yargs) {
    return yargs.command('component <name>', 'Deploys a new component (or updates an existing component) to the Sitecore server when using Sitecore-first development. `jss deploy component --help` for options.', args, handler);
}
exports.default = builder;
/**
 * @param {Argv} yargs
 */
function args(yargs) {
    return deploy_template_1.args(yargs)
        .option('allowedPlaceholders', {
        requiresArg: false,
        type: 'array',
        describe: 'The placeholder name(s) to allow the rendering to be placed in. For example --allowedPlaceholders Foo or --allowedPlaceholders Foo Bar',
    })
        .option('exposesPlaceholders', {
        requiresArg: false,
        type: 'array',
        describe: 'The names of any placeholders that are exposed on this component (to place other components in). This will cause placeholder settings items to be generated for them.',
    });
}
exports.args = args;
/**
 * @param {any} argv
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function handler(argv) {
    return __awaiter(this, void 0, void 0, function () {
        var fields, definition, manifestSource;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    fields = [];
                    if (argv.fields && argv.fields.length > 0) {
                        argv.fields.forEach(function (field) {
                            var splitField = field.split(':');
                            var fieldType = splitField.length === 1 ? 'Single-Line Text' : splitField[1];
                            fields.push({ name: splitField[0], type: fieldType });
                        });
                    }
                    definition = {
                        name: argv.name,
                        displayName: argv.displayName,
                        icon: argv.icon,
                        fields: fields.length === 0 ? null : fields,
                        placeholders: argv.exposesPlaceholders,
                        allowedPlaceholders: argv.allowedPlaceholders,
                    };
                    manifestSource = "module.exports = {\n  default: function(manifest) {\n    manifest.addComponent(" + JSON.stringify(definition, null, 2) + ");\n  }\n}";
                    if (!argv.skipDeploy) return [3 /*break*/, 1];
                    console.log(manifestSource);
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, micro_manifest_1.default(argv, manifestSource)];
                case 2:
                    _a.sent();
                    console.log("\n" + chalk_1.default.green('Your component has been created (or updated)!') + "\n\n" + chalk_1.default.green('NEXT STEPS FOR NEW COMPONENTS') + "\n\n* Create a component in your JSS application called " + chalk_1.default.green(argv.name) + " (" + chalk_1.default.green("jss scaffold " + argv.name) + ")\n* Deploy your app with the new component to Sitecore (" + chalk_1.default.green('jss deploy:watch') + " or " + chalk_1.default.green('jss deploy files') + ")\n* Add the new component to a route item using the Experience Editor\n");
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
