// @sitecore-jss/sitecore-jss-cli/dist/scripts/index.js
"use strict";
// commands available when running from an app location (with CLI installed to local node_modules)
// e.g. setup, deploy
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clean = exports.manifest = exports.pkg = exports.deploy = exports.setup = exports.environment = exports.elephant = exports.easterEgg = void 0;
var easterEgg = __importStar(require("./easter-egg"));
exports.easterEgg = easterEgg;
var elephant = __importStar(require("./elephant"));
exports.elephant = elephant;
var environment = __importStar(require("./environment"));
exports.environment = environment;
var clean = __importStar(require("./clean"));
exports.clean = clean;
var deploy = __importStar(require("./deploy"));
exports.deploy = deploy;
var manifest = __importStar(require("./manifest"));
exports.manifest = manifest;
var pkg = __importStar(require("./package"));
exports.pkg = pkg;
var setup = __importStar(require("./setup"));
exports.setup = setup;
